chore(docs): Include import case to MDX setup (#19286)
gatsbyjs/gatsby: docs/docs/mdx/programmatically-creating-pages.md

`MDXRenderer` which is used by `gatsby-plugin-mdx` that will be used to render any
programmatically accessed MDX content.

> For now, to update imports withing .mdx files, you should re-run your gatsby development environment. Otherwise, it will raise the ReferenceError. To import things dynamically, we will use the `MDXProvider` component and provide all common components we want to use, like `Link`, for example.

Reviewer: as i understand the MDXProvider docs then you need only supply a link component if you like to replace a default react component?

Reviewer: maybe with this rewording?

```suggestion
For now, to update imports withing .mdx files, you should re-run your gatsby development environment. Otherwise, it will raise the ReferenceError. To import things dynamically, we will use the `MDXProvider` component and provide all common components we want to use. In the example it replace the `a` tag with the Gatsby `Link` instead of the default React Link element.
```

> For now, to update imports withing .mdx files, you should re-run your gatsby development environment. Otherwise, it will raise the ReferenceError. To import things dynamically, we will use the `MDXProvider` component and provide all common components we want to use, like `Link`, for example.

Reviewer 2: ```suggestion
For now, to update imports within .mdx files, you should re-run your gatsby development environment. Otherwise, it will raise a ReferenceError. To import things dynamically, you can use the `MDXProvider` component and provide it all the common components you'll be using, such as `Link`.
```

Author: Thanks a lot!

First, create a component that accepts the queried MDX data (which will be
added in the next step).

```javascript:title=src/components/posts-page-layout.js
import React from "react"
import { graphql } from "gatsby"
import { MDXProvider } from "@mdx-js/react"
import { MDXRenderer } from "gatsby-plugin-mdx"
import { Link } from "gatsby"

export default function PageTemplate({ data: { mdx } }) {
  return (
    <div>
      <h1>{mdx.frontmatter.title}</h1>

>       <MDXProvider components={{ Link }}>

Reviewer: should this not be like this?

```suggestion
      <MDXProvider components={{ a: Link }}>
```

Author: I'm not sure, but I think no. Because I use both `<Link>` and `<a>` tags for internal and external links respectively, as it recommended [here](https://www.gatsbyjs.org/docs/gatsby-link/#replace-a-tags-with-the-link-tag-for-local-links). And I believe it is better to keep original names of the imported components. The idea is taken from `gatsby-plugin-mdx` [shortcodes](https://www.gatsbyjs.org/packages/gatsby-plugin-mdx/#shortcodes) section.

On the other hand, in [this section](https://www.gatsbyjs.org/docs/mdx/customizing-components/), I see how to customize default HTML elements. But that is not precisely the case I try to describe with this PR. However, it still shows the import mechanism. 

What do you think?

Reviewer: ok. I understand it now why you include the link. 

Would it be less confusing to use also a second other component for the example? The it is more clear I think. Or include links to the example in the shortcodes section?

What do you mean?

Author: Yes, I agree with you. I'll make that soon.

Reviewer: one thing is confusing me here: if the default setting to exchange all `a`-tags with Gaysby `Link` why he should include it then with `MDXProvider`?

Author: Here is my case https://redfish-project.gq/docs

I use `<a>` tag to link external services or libraries like AWS, Apollo, etc.
But `<Link>` tag to navigate a user to move toward other sections of the documentation.

My goal is to provide a mechanism on how MDXProvider can import components globally and explain the difference between different import approaches. So I presume a reader will decide what to do with that.

Author: > ok. I understand it now why you include the link.
> 
> Would it be less confusing to use also a second other component for the example? The it is more clear I think. Or include links to the example in the shortcodes section?

I've implemented this suggestion except for the second component idea. Not sure which one to include. It could be redundant. I wrote a comment instead of that.

`MDXRenderer` which is used by `gatsby-plugin-mdx` that will be used to render any
programmatically accessed MDX content.

> For now, to update imports within .mdx files, you should re-run your gatsby development environment. Otherwise, it will raise a ReferenceError. To import things dynamically, you can use the `MDXProvider` component and provide it all the common components you'll be using, such as `Link`.

Reviewer 3: ```suggestion
For now, to update imports within .mdx files, you should rerun your Gatsby development environment. Otherwise, it will raise a `ReferenceError`. To import things dynamically, you can use the `MDXProvider` component and provide it all the common components you'll be using, such as `Link`.
```
